sup3r/bias/qdm.py

```diff
@@ -495,3 +505,314 @@ def run(self,
+       hydrological simulations of climate change. Journal of Hydrometeorology,
+       16(6), 2421-2442.
+    # Todo:
+    #   - Identify Z_gf. (0.01 mm) Also have to save on output params
+    #   - Estimate K = <x> / <hat{x}>
+    """
+
+    def _init_out(self):
+        super()._init_out()
+
+        shape = (*self.bias_gid_raster.shape, 1)
+        self.out[f'{self.base_dset}_zero_rate'] = np.full(shape,
+                                                          np.nan,
+                                                          np.float32)
+        shape = (*self.bias_gid_raster.shape, 12)
+        self.out[f'{self.bias_feature}_mean_mh'] = np.full(shape,
+                                                           np.nan,
+                                                           np.float32)
+        self.out[f'{self.bias_feature}_mean_mf'] = np.full(shape,
+                                                           np.nan,
+                                                           np.float32)
+
+    # pylint: disable=W0613
+    @classmethod
+    def _run_single(
+        cls,
+        bias_data,
+        bias_fut_data,
+        base_fps,
+        bias_feature,
+        base_dset,
+        base_gid,
+        base_handler,
+        daily_reduction,
+        decimals,
+        sampling,
+        n_samples,
+        log_base,
+        *,
+        bias_ti,
+        bias_fut_ti,
+        zero_rate_threshold,
+        base_dh_inst=None,
+
+    ):
+        """Estimate probability distributions at a single site
+
+        ATTENTION: This should be refactored. There is too much redundancy in
+        the code. Let's make it work first, and optimize later.
+        """
+        base_data, base_ti = cls.get_base_data(
+            base_fps,
+            base_dset,
+            base_gid,
+            base_handler,
+            daily_reduction=daily_reduction,
+            decimals=decimals,
+            base_dh_inst=base_dh_inst,
+        )
+
+        out = cls.get_qdm_params(
+            bias_data,
+            bias_fut_data,
+            base_data,
+            bias_feature,
+            base_dset,
+            sampling,
+            n_samples,
+            log_base,
+        )
+
+        out[f'{base_dset}_zero_rate'] = cls.zero_precipitation_rate(
+            base_data,
+            zero_rate_threshold,
+        )
+
+        # Let's save the means for mh and mf instead of the `x` ratio. It
+        # seems that we should be able to simplify the mh component from
+        # the `K` coefficient.
+        # TODO: For now it is monthly but later it will be modified to a
+        # generic time window.
+        mh = np.full(12, np.nan, np.float32)
+        mf = np.full(12, np.nan, np.float32)
+        for m in range(12):
+            mh[m] = bias_data[bias_ti.month == (m + 1)].mean()
+            mf[m] = bias_fut_data[bias_fut_ti.month == (m + 1)].mean()
+        out[f'{bias_feature}_mean_mh'] = mh
+        out[f'{bias_feature}_mean_mf'] = mf
```

Comment:
Dont we also need the mean value of the modeled future data after bias correction to calculate x^?

Reply:
We are doing something else now. I'll close this one. Please feel free to reopen if you still want to discuss more on this instead of the new one.